## **Overview**

In this assignment, you will perform a personalized data analysis using **NumPy** and **Pandas** on a real-world dataset derived from the *World Development Indicators (1990–2020)*.

Your dataset will be a unique random sample of the same master file, generated using your **student ID**.

You will demonstrate your ability to:

- Perform numerical computation using **NumPy**
- Manipulate and summarize data using **Pandas**
- Draw reasoned conclusions from real-world, incomplete data

---

## **1. Dataset and Notebook**

You are provided with a master dataset named **`WDI_dataset.csv`** containing the following columns:

| Column | Description |
|--------|-------------|
| **Country** | Country name |
| **Year** | Year of observation (1990–2020) |
| **LifeExpectancy** | Life expectancy at birth (years) |
| **GDPperCapita** | GDP per capita (USD) |
| **Population** | Total population |
| **SchoolEnrollment** | Gross school enrollment (%) |
| **InternetUsers** | % of population using the Internet |
| **HealthExpenditure** | Health expenditure (% of GDP) |
| **LiteracyRate** | Literacy rate (%) |
| **GDPGrowth** | Annual GDP growth rate (%) |

---

### **Notebook Requirement**

You must complete this assignment in a **Jupyter Notebook (`.ipynb`)** file.

> **Only `.ipynb` submissions will be accepted**; `.py` or other formats will **not be marked**.


In [108]:
#Sampling Code
import pandas as pd

# Load master dataset
df_wdi_full = pd.read_csv("wdi_dataset.csv")

# Define your student ID
student_id = "nf1023374"  # Replace with your own ID
SAMPLE_FRACTION = 0.15
SEED_DIGITS = 3 

# Derive random seed
seed = int(str(student_id)[-SEED_DIGITS:])

# Create your sample
df_sample = df_wdi_full.sample(frac=SAMPLE_FRACTION, random_state=seed).reset_index(drop=True)

# Save the sample
filename = f"student_sample_{student_id}.csv"
df_sample.to_csv(filename, index=False)

print(f"Your sample ({len(df_sample)} rows) has been saved as {filename}")
print(f"Seed used: {seed}")


Your sample (1232 rows) has been saved as student_sample_nf1023374.csv
Seed used: 374



## **TASK 1 – Dataset Overview**

Perform an initial exploration of the dataset to understand its structure.

### **Instructions**
1. **Display the number of rows and columns** in the dataset.
2. **List all column names** and their **data types**.
3. Show the **range of years** present in the dataset.
4. Report the **number of unique countries**.

In [109]:
print(f"{df_sample.info()}")
df_sample.head() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1232 entries, 0 to 1231
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            1232 non-null   object 
 1   Year               1232 non-null   int64  
 2   LifeExpectancy     1232 non-null   float64
 3   GDPperCapita       1187 non-null   float64
 4   Population         1232 non-null   float64
 5   SchoolEnrollment   850 non-null    float64
 6   InternetUsers      914 non-null    float64
 7   HealthExpenditure  730 non-null    float64
 8   LiteracyRate       306 non-null    float64
 9   GDPGrowth          1168 non-null   float64
dtypes: float64(8), int64(1), object(1)
memory usage: 96.4+ KB
None


,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
0,Brazil,1996,68.108000,5179.130520,1.642025e+08,NaN,0.451000,NaN,NaN,2.208864
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733
3,Belgium,1998,77.473171,25338.443293,1.020301e+07,NaN,7.890000,NaN,NaN,1.961832
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000


In [110]:
print(f"the sample has {df_sample.shape} rows and columns")
print(f"colums names are: {df_sample.columns.tolist()}")
print(f"the range of years is: {df_sample['Year'].min()} - {df_sample['Year'].max()}")
print(f"num of unique countries: {df_sample['Country'].nunique()}")

the sample has (1232, 10) rows and columns
colums names are: ['Country', 'Year', 'LifeExpectancy', 'GDPperCapita', 'Population', 'SchoolEnrollment', 'InternetUsers', 'HealthExpenditure', 'LiteracyRate', 'GDPGrowth']
the range of years is: 1990 - 2020
num of unique countries: 263


Con el método `head()` de **pandas** podemos visualizar las primeras *n* filas de un DataFrame.  
Por defecto, `n = 5`.

```python
df_sample.head()        # Muestra las primeras 5 filas
df_sample.head(10)      # Muestra las primeras 10 filas
```
Para conocer cuántas filas y columnas tiene un DataFrame existen varias opciones.
Por ejemplo:
```python
len(df_sample.columns)   # Cantidad de columnas
len(df_sample)           # Cantidad de filas

```
Sin embargo, la forma más eficiente es usar df_sample.shape, que devuelve una tupla:
```python 
df_sample.shape
```
Esta tupla tiene la siguiente estructura:

El primer elemento representa el número de filas.
El segundo elemento representa el número de columnas.

## **TASK 2 – Filtering Data**
1. Filter data for **years ≥ 2000** and count rows.
2. Find observations where **LifeExpectancy > 75** and **GDPperCapita > 30,000**.
3. List the **first 10** countries meeting those conditions.


In [111]:
df_sample[df_sample["Year"]>=2000]

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000
5,Malaysia,2018,75.734000,10901.801995,3.291097e+07,95.776154,81.201000,409.847107,94.849998,4.843087
7,American Samoa,2011,72.662000,10383.459331,5.489500e+04,NaN,NaN,NaN,NaN,0.000000
...,...,...,...,...,...,...,...,...,...,...
1225,"Gambia, The",2011,61.561000,710.608411,1.983784e+06,72.190663,10.870300,31.051489,NaN,-8.130444
1226,East Asia & Pacific,2013,75.334773,9522.667498,2.267072e+09,104.620399,43.200000,603.489925,94.617699,5.273683
1227,Antigua and Barbuda,2016,76.968000,16556.855180,8.996900e+04,107.432655,70.974300,817.888916,NaN,4.120368
1229,Slovenia,2018,81.378049,25888.048107,2.073894e+06,99.463097,79.750000,2169.578857,NaN,4.379083


By applying this filter (`df_sample[df_sample["Year"] >= 2000]`), we reduce the dataset to **811 rows**, which means that **more than half of all records correspond to years after 2000**.  
This allows us to focus the analysis on more recent and relevant data.

In [112]:
df_sample[(df_sample["LifeExpectancy"] > 75) & (df_sample["GDPperCapita"] >= 30000)]

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
10,OECD members,2016,79.891423,36235.998825,1.340142e+09,102.840218,NaN,4445.736994,NaN,1.843324
24,New Caledonia,2013,76.867000,36671.156675,2.768220e+05,NaN,66.0000,NaN,NaN,NaN
25,Isle of Man,2015,80.267000,84753.518742,8.359300e+04,NaN,NaN,NaN,NaN,-0.385867
28,United Arab Emirates,2018,82.449000,45689.857005,9.346701e+06,105.675358,98.4500,1872.584229,NaN,1.313914
35,Qatar,2007,77.741000,65954.341953,1.208595e+06,98.263962,37.0000,1301.496826,93.080002,17.985657
...,...,...,...,...,...,...,...,...,...,...
1199,New Zealand,2013,81.404878,43000.464685,4.442100e+06,98.876289,82.7800,4012.083740,NaN,2.589542
1200,Switzerland,2019,83.904878,84121.931030,8.575280e+06,101.593048,93.1461,9646.913086,NaN,1.141982
1203,Luxembourg,2017,82.095122,110193.213797,5.963360e+05,96.735199,97.3630,5713.118652,NaN,1.317187
1220,Channel Islands,2014,81.600000,64700.742392,1.627260e+05,NaN,NaN,NaN,NaN,2.427102


Using the filter  
`df_sample[(df_sample["LifeExpectancy"] > 75) & (df_sample["GDPperCapita"] >= 30000)]`,  
we can observe that it is **quite rare** for a country to have both a **very high life expectancy** and a **high GDP per capita**.

Only **137 rows out of more than 1200 records** meet these conditions, showing that achieving both indicators simultaneously is relatively uncommon at a global level.

In [113]:
df_sample[(df_sample["Year"] > 2000) & (df_sample["LifeExpectancy"]>75) & (df_sample["GDPperCapita"]>30000)].head(10)

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth
10,OECD members,2016,79.891423,36235.998825,1.340142e+09,102.840218,NaN,4445.736994,NaN,1.843324
24,New Caledonia,2013,76.867000,36671.156675,2.768220e+05,NaN,66.0000,NaN,NaN,NaN
25,Isle of Man,2015,80.267000,84753.518742,8.359300e+04,NaN,NaN,NaN,NaN,-0.385867
28,United Arab Emirates,2018,82.449000,45689.857005,9.346701e+06,105.675358,98.4500,1872.584229,NaN,1.313914
35,Qatar,2007,77.741000,65954.341953,1.208595e+06,98.263962,37.0000,1301.496826,93.080002,17.985657
56,Austria,2006,79.880488,40382.207830,8.268641e+06,102.310898,63.6000,3882.981934,NaN,3.269336
58,"Macao SAR, China",2016,83.275610,69074.036305,6.525000e+05,101.287356,81.6430,NaN,96.540001,-0.679184
64,New Zealand,2020,82.256098,41850.920328,5.090200e+06,97.536530,95.2342,4214.571777,NaN,-0.307247
66,New Zealand,2018,81.707317,43257.076924,4.900600e+06,98.552971,89.0000,3907.599121,NaN,3.470582
73,Finland,2008,79.568293,53769.472624,5.313399e+06,99.719872,83.6700,4486.337891,NaN,0.784431


By applying the filter  
`df_sample[(df_sample["Year"] > 2000) & (df_sample["LifeExpectancy"] > 75) & (df_sample["GDPperCapita"] > 30000)]`,  
we can see that **all countries meeting these criteria are highly developed nations**.

This confirms that achieving both a **high life expectancy** and a **high GDP per capita** in the 21st century is strongly associated with countries that have advanced economies, strong healthcare systems, and high standards of living.

## **TASK 3 – Grouping and Aggregation**
1.	Compute average **life expectancy and average GDP per capita by country.**
2.	Display the **top 5 countries** for each measure.
3.	Compute the average **GDP growth rate per decade (1990s, 2000s, 2010s).**


In [114]:
df_sample.groupby("Country").agg({"LifeExpectancy": "mean", "GDPperCapita": "mean"}).\
sort_values(by = "LifeExpectancy", ascending=False).head(5)

,LifeExpectancy,GDPperCapita
Country,,
"Hong Kong SAR, China",83.442520,35494.032007
Monaco,83.265500,134725.571958
Japan,82.465915,37125.894197
Andorra,82.418750,27042.962293
San Marino,82.149143,50449.164277


In [115]:
# Different way to do it
df_sample[["LifeExpectancy", "GDPperCapita","Country" ]].groupby("Country").mean()\
    .sort_values(by = "GDPperCapita", ascending= False).head(5)

,LifeExpectancy,GDPperCapita
Country,,
Monaco,83.265500,134725.571958
Liechtenstein,81.275171,127238.104931
Luxembourg,78.773557,81148.352948
Cayman Islands,75.305800,75107.662301
Bermuda,78.034500,73284.730170


Using the grouped dataset:

we observe that countries with the highest average life expectancy also tend to have high GDP per capita.

This suggests a possible positive correlation between economic prosperity and population health:
countries with greater financial resources often invest more in healthcare, education, and quality of life, leading to higher life expectancy.

In [116]:
df_sample["Decade"] = df_sample["Year"] // 10 * 10
df_sample.groupby("Decade").agg({"GDPperCapita": "mean"}).head(5)

,GDPperCapita
Decade,
1990,6747.593161
2000,11554.896114
2010,17574.327733
2020,12058.133822


we first create a new column called `Decade` using:

`df_sample["Decade"] = df_sample["Year"] // 10 * 10`

This converts each year into its corresponding decade (e.g., 1994 → 1990, 2007 → 2000).  
Then we group the dataset by this new column and calculate the **average GDP per capita** for each decade:

`df_sample.groupby("Decade").agg({"GDPperCapita": "mean"}).head(5)`

With this, we can observe how **GDP per capita changes over time**


## **TASK 4 – Sorting and Extremes**
1.	Find the observation with the highest life expectancy and lowest GDP per capita.
1.	Display the top 10 records by life expectancy.


In [117]:
df_sample["GDPperCapita"].idxmin()

1146

In [118]:
df_sample.loc[df_sample["LifeExpectancy"].idxmax()]


Country              Hong Kong SAR, China
Year                                 2020
LifeExpectancy                  85.496341
GDPperCapita                 46109.229995
Population                      7481000.0
SchoolEnrollment               105.128763
InternetUsers                     92.4131
HealthExpenditure                     NaN
LiteracyRate                          NaN
GDPGrowth                       -6.544783
Decade                               2020
Name: 355, dtype: object

In [119]:
df_sample.loc[df_sample["GDPperCapita"].idxmin()]

Country                    Iraq
Year                       1993
LifeExpectancy            65.31
GDPperCapita          53.743784
Population           19201195.0
SchoolEnrollment      91.135147
InternetUsers               NaN
HealthExpenditure           NaN
LiteracyRate                NaN
GDPGrowth             30.289829
Decade                     1990
Name: 1146, dtype: object

In [120]:
df_sample.loc[
    [df_sample["GDPperCapita"].idxmin(), df_sample["LifeExpectancy"].idxmax()]
]


,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade
1146,Iraq,1993,65.310000,53.743784,19201195.0,91.135147,NaN,NaN,NaN,30.289829,1990
355,"Hong Kong SAR, China",2020,85.496341,46109.229995,7481000.0,105.128763,92.4131,NaN,NaN,-6.544783,2020


Here we use the functions **`idxmin()`** and **`idxmax()`** to identify the exact rows where the dataset reaches:

- the **minimum GDP per Capita**, and  
- the **maximum Life Expectancy**.

The code:

`df_sample.loc[[df_sample["GDPperCapita"].idxmin(), df_sample["LifeExpectancy"].idxmax()]]`

returns the full records for both cases, allowing us to clearly see **which countries and years correspond to the lowest economic value and the highest health outcome** in the dataset.

In [121]:
df_sample.sort_values(by="LifeExpectancy", ascending=False).head(10)

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade
355,"Hong Kong SAR, China",2020,85.496341,46109.229995,7481000.0,105.128763,92.4131,NaN,NaN,-6.544783,2020
204,Monaco,2013,85.056000,184940.663805,35447.0,NaN,90.7000,8455.655273,NaN,9.577098,2010
649,San Marino,2017,84.933000,45192.027977,33825.0,NaN,60.1823,3672.303223,NaN,0.258265,2010
1158,Monaco,2010,84.644000,161853.920622,33163.0,NaN,75.0000,8016.669922,NaN,2.065220,2010
452,San Marino,2014,84.394000,51260.493835,32655.0,NaN,NaN,4120.026855,NaN,-0.702371,2010
755,San Marino,2013,84.175000,50807.828914,33041.0,NaN,NaN,4113.466797,NaN,-0.783972,2010
266,Liechtenstein,2019,84.160976,166907.839305,38563.0,103.830009,93.1461,NaN,NaN,NaN,2010
664,Andorra,2009,84.100000,43975.018367,83888.0,85.024712,78.5300,2663.374023,NaN,-5.302847,2000
757,Japan,2017,84.099756,38834.052934,126972000.0,102.254829,91.7265,4135.243164,NaN,1.675332,2010
1200,Switzerland,2019,83.904878,84121.931030,8575280.0,101.593048,93.1461,9646.913086,NaN,1.141982,2010


If we sort the dataset by **LifeExpectancy**, we can observe that several countries appear repeatedly across different years.  
These countries consistently rank at the top, and they are typically **high-income, highly developed countries** with strong healthcare systems and high living standards.

## **TASK 5 – Counting and Distributions**
1.	Count how many observations exist per year and decade.
2.	Display the number of unique countries per decade.


In [122]:
# 1.	Count how many observations exist per year and decade.
df_sample.groupby("Year").size()


Year
1990    53
1991    35
1992    48
1993    39
1994    43
1995    31
1996    38
1997    47
1998    45
1999    42
2000    39
2001    32
2002    38
2003    38
2004    37
2005    35
2006    36
2007    42
2008    43
2009    35
2010    29
2011    35
2012    31
2013    38
2014    48
2015    43
2016    41
2017    40
2018    39
2019    45
2020    47
dtype: int64

In [123]:
df_sample.groupby("Decade").size()

Decade
1990    421
2000    375
2010    389
2020     47
dtype: int64

The dataset is fairly **balanced across the 1990s, 2000s, and 2010s**, showing a consistent amount of data for each period.  
The **2020s have fewer records** simply because the decade is just beginning and the dataset only covers the first years.

In [124]:
NUM_unique_countries_per_decade = df_sample.groupby("Decade")["Country"].nunique()
unique_countries_per_decade = df_sample.groupby("Decade")["Country"].unique()
print(f"Number of unique countries per decade:\n{NUM_unique_countries_per_decade}\
n\n-----------------------\nUnique countries per decade:\n{unique_countries_per_decade}")

Number of unique countries per decade:
Decade
1990    218
2000    204
2010    206
2020     47
Name: Country, dtype: int64n
-----------------------
Unique countries per decade:
Decade
1990    [Brazil, Belgium, Guinea-Bissau, Malta, Norway...
2000    [India, Ghana, Slovenia, Barbados, Fiji, Jamai...
2010    [Mauritania, Malaysia, American Samoa, Least d...
2020    [Myanmar, Lithuania, New Zealand, St. Vincent ...
Name: Country, dtype: object


In [125]:
_countris_2020 = unique_countries_per_decade[2020]
_countris_2020

array(['Myanmar', 'Lithuania', 'New Zealand',
       'St. Vincent and the Grenadines', 'Malawi', 'Congo, Dem. Rep.',
       'Belgium', 'Latin America & the Caribbean (IDA & IBRD countries)',
       'Bhutan', 'Fragile and conflict affected situations', 'Iraq',
       'Iran, Islamic Rep.', 'Zambia', 'Singapore',
       'Hong Kong SAR, China', 'Finland', 'Barbados', 'Tunisia',
       'Latin America & Caribbean (excluding high income)', 'Chile',
       'Brazil', 'Solomon Islands', 'Slovak Republic', 'Timor-Leste',
       'Trinidad and Tobago', 'Sub-Saharan Africa', 'Honduras',
       'Egypt, Arab Rep.', 'Russian Federation', 'Lebanon', 'Niger',
       'Chad', 'Least developed countries: UN classification', 'World',
       'Latvia', 'East Asia & Pacific (IDA & IBRD countries)',
       'Brunei Darussalam',
       'Middle East, North Africa, Afghanistan & Pakistan (IDA & IBRD)',
       'Upper middle income', 'Bulgaria', 'Antigua and Barbuda', 'Guyana',
       'Sierra Leone', 'United Arab Emir

## **TASK 6 – Creating New Columns** 
Create:
1.	PopulationMillions = Population / 1_000_000
2.	GDPCategory = "Low" / "Medium" / "High" based on:
3.	Low: below median GDP
4.	Medium: between median and 75th percentile
5.	High: above 75th percentile

Then count how many rows fall into each category.


In [126]:
df_sample["PopulationMillions"] = df_sample["Population"] / 1_000_000

# Method 1: Using built-in pandas functions pd.qcut
df_sample["GDPCategory"] = pd.qcut(df_sample["GDPperCapita"], q=[0, 0.5, 0.75, 1], labels=["Low", "Medium", "High"])
df_sample

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory
0,Brazil,1996,68.108000,5179.130520,1.642025e+08,NaN,0.451000,NaN,NaN,2.208864,1990,164.202549,Medium
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546,2010,4.210600,Low
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733,2000,1172.878890,Low
3,Belgium,1998,77.473171,25338.443293,1.020301e+07,NaN,7.890000,NaN,NaN,1.961832,1990,10.203008,High
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000,2000,19.637087,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1227,Antigua and Barbuda,2016,76.968000,16556.855180,8.996900e+04,107.432655,70.974300,817.888916,NaN,4.120368,2010,0.089969,High
1228,Cote d'Ivoire,1992,52.429000,844.902224,1.320031e+07,63.168678,NaN,NaN,NaN,-0.244561,1990,13.200310,Low
1229,Slovenia,2018,81.378049,25888.048107,2.073894e+06,99.463097,79.750000,2169.578857,NaN,4.379083,2010,2.073894,High
1230,"Gambia, The",1999,56.466000,575.459893,1.415779e+06,71.793800,0.716000,NaN,NaN,6.399999,1990,1.415779,Low


In [127]:
# Method 2: Using a custom function
def gdp_category(row):
    if row["GDPperCapita"] < df_sample["GDPperCapita"].median():
        return "Low"
    elif row["GDPperCapita"] <= df_sample["GDPperCapita"].quantile(0.75):
        return "Medium"
    else:
        return "High"

df_sample["GDPCategory_custom"] = df_sample.apply(gdp_category, axis=1)
df_sample

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory,GDPCategory_custom
0,Brazil,1996,68.108000,5179.130520,1.642025e+08,NaN,0.451000,NaN,NaN,2.208864,1990,164.202549,Medium,Medium
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546,2010,4.210600,Low,Low
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733,2000,1172.878890,Low,Low
3,Belgium,1998,77.473171,25338.443293,1.020301e+07,NaN,7.890000,NaN,NaN,1.961832,1990,10.203008,High,High
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000,2000,19.637087,Low,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1227,Antigua and Barbuda,2016,76.968000,16556.855180,8.996900e+04,107.432655,70.974300,817.888916,NaN,4.120368,2010,0.089969,High,High
1228,Cote d'Ivoire,1992,52.429000,844.902224,1.320031e+07,63.168678,NaN,NaN,NaN,-0.244561,1990,13.200310,Low,Low
1229,Slovenia,2018,81.378049,25888.048107,2.073894e+06,99.463097,79.750000,2169.578857,NaN,4.379083,2010,2.073894,High,High
1230,"Gambia, The",1999,56.466000,575.459893,1.415779e+06,71.793800,0.716000,NaN,NaN,6.399999,1990,1.415779,Low,Low


In [128]:
# Method 3: Using a lambda function
df_sample["GDPCategory_lambda"] = df_sample.apply(
    lambda row: "Low" if row["GDPperCapita"] < df_sample["GDPperCapita"].median() 
    else "Mediun" if row["GDPperCapita"] <= df_sample["GDPperCapita"].quantile(0.75) 
    else "High", axis=1) 
df_sample

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory,GDPCategory_custom,GDPCategory_lambda
0,Brazil,1996,68.108000,5179.130520,1.642025e+08,NaN,0.451000,NaN,NaN,2.208864,1990,164.202549,Medium,Medium,Mediun
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546,2010,4.210600,Low,Low,Low
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733,2000,1172.878890,Low,Low,Low
3,Belgium,1998,77.473171,25338.443293,1.020301e+07,NaN,7.890000,NaN,NaN,1.961832,1990,10.203008,High,High,High
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000,2000,19.637087,Low,Low,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1227,Antigua and Barbuda,2016,76.968000,16556.855180,8.996900e+04,107.432655,70.974300,817.888916,NaN,4.120368,2010,0.089969,High,High,High
1228,Cote d'Ivoire,1992,52.429000,844.902224,1.320031e+07,63.168678,NaN,NaN,NaN,-0.244561,1990,13.200310,Low,Low,Low
1229,Slovenia,2018,81.378049,25888.048107,2.073894e+06,99.463097,79.750000,2169.578857,NaN,4.379083,2010,2.073894,High,High,High
1230,"Gambia, The",1999,56.466000,575.459893,1.415779e+06,71.793800,0.716000,NaN,NaN,6.399999,1990,1.415779,Low,Low,Low


np.float64(3700.959567751443)

In [130]:
df_sample[[ "GDPperCapita","GDPCategory", "GDPCategory_lambda"]].head(10)

,GDPperCapita,GDPCategory,GDPCategory_lambda
0,5179.130520,Medium,Mediun
1,1615.004013,Low,Low
2,801.668354,Low,Low
3,25338.443293,High,High
4,253.746936,Low,Low
5,10901.801995,Medium,Mediun
6,574.321106,Low,Low
7,10383.459331,Medium,Mediun
8,1104.186332,Low,Low
9,23343.362897,High,High


In [151]:
df_sample["GDPCategory_custom"].value_counts()

GDPCategory_custom
Low       593
High      342
Medium    297
Name: count, dtype: int64

Half of the dataset fis the median and is labeled as “Low,” while the Medium and High groups are nearly the same size because they follow the 50–75% and 75–100% quantile ranges.  
`qcut` handles boundaries and ties automatically, keeping the groups balanced, unlike the lambda and custom versions that uses fixed rules. I preferred using `qcut` to avoid issues with balancing and edge cases.

## **TASK 7 – Missing Data Analysis**
1.	Compute and display missing values (%) for all numeric columns.
2.	Identify which variable has the most missing data.
3.	Count rows with no missing values vs at least one missing.


In [132]:
missing_values = df_sample.isnull().mean() * 100
missing_values

Country                0.000000
Year                   0.000000
LifeExpectancy         0.000000
GDPperCapita           3.652597
Population             0.000000
SchoolEnrollment      31.006494
InternetUsers         25.811688
HealthExpenditure     40.746753
LiteracyRate          75.162338
GDPGrowth              5.194805
Decade                 0.000000
PopulationMillions     0.000000
GDPCategory            3.652597
GDPCategory_custom     0.000000
GDPCategory_lambda     0.000000
dtype: float64

In [133]:
missing_values_percent = df_sample.isnull().sum() / df_sample.shape[0] * 100
missing_values_percent

Country                0.000000
Year                   0.000000
LifeExpectancy         0.000000
GDPperCapita           3.652597
Population             0.000000
SchoolEnrollment      31.006494
InternetUsers         25.811688
HealthExpenditure     40.746753
LiteracyRate          75.162338
GDPGrowth              5.194805
Decade                 0.000000
PopulationMillions     0.000000
GDPCategory            3.652597
GDPCategory_custom     0.000000
GDPCategory_lambda     0.000000
dtype: float64

In [134]:
not_missing = df_sample.notnull().all(axis=1).sum() 
missing = df_sample.isnull().any(axis=1).sum() 
print(f"Rows with no missing values: {not_missing}") 
print(f"Rows with at least one missing value: {missing}")

Rows with no missing values: 170
Rows with at least one missing value: 1062



Out of the entire dataset, **170 rows have no missing values**, while **1,062 rows contain at least one missing value**.  
This means the dataset has a **high proportion of incomplete records**, which is important to consider for cleaning, filtering, or imputation before deeper analysis.


## **TASK 8 – Conditional Selections** 
1. Select all rows where LifeExpectancy > 80.
2.	Select rows where HealthExpenditure > 8% and GDPGrowth > 3%.
3.	Compare the average literacy rate between these two subsets.


In [135]:
life_exp = df_sample[df_sample["LifeExpectancy"] > 80]
helth_exp = df_sample[(df_sample["HealthExpenditure"] > 8) & (df_sample["GDPGrowth"] > 3)]
life_exp

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory,GDPCategory_custom,GDPCategory_lambda
9,Portugal,2019,81.675610,23343.362897,10286263.0,104.897682,75.3464,2221.731201,NaN,2.745685,2010,10.286263,High,High,High
25,Isle of Man,2015,80.267000,84753.518742,83593.0,NaN,NaN,NaN,NaN,-0.385867,2010,0.083593,High,High,High
28,United Arab Emirates,2018,82.449000,45689.857005,9346701.0,105.675358,98.4500,1872.584229,NaN,1.313914,2010,9.346701,High,High,High
58,"Macao SAR, China",2016,83.275610,69074.036305,652500.0,101.287356,81.6430,NaN,96.540001,-0.679184,2010,0.652500,High,High,High
64,New Zealand,2020,82.256098,41850.920328,5090200.0,97.536530,95.2342,4214.571777,NaN,-0.307247,2020,5.090200,High,High,High
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1200,Switzerland,2019,83.904878,84121.931030,8575280.0,101.593048,93.1461,9646.913086,NaN,1.141982,2010,8.575280,High,High,High
1203,Luxembourg,2017,82.095122,110193.213797,596336.0,96.735199,97.3630,5713.118652,NaN,1.317187,2010,0.596336,High,High,High
1220,Channel Islands,2014,81.600000,64700.742392,162726.0,NaN,NaN,NaN,NaN,2.427102,2010,0.162726,High,High,High
1223,Singapore,2013,82.246341,56967.425794,5399162.0,NaN,80.9021,2102.970215,96.550003,4.817631,2010,5.399162,High,High,High


In [136]:
helth_exp

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory,GDPCategory_custom,GDPCategory_lambda
1,Mauritania,2017,67.088000,1615.004013,4.210600e+06,87.817309,21.017200,55.224464,NaN,6.270546,2010,4.210600,Low,Low,Low
2,India,2006,65.373000,801.668354,1.172879e+09,NaN,2.805500,29.376009,62.750000,8.060733,2000,1172.878890,Low,Low,Low
4,Ghana,2000,58.139000,253.746936,1.963709e+07,79.940376,0.153615,10.451775,57.900002,3.700000,2000,19.637087,Low,Low,Low
5,Malaysia,2018,75.734000,10901.801995,3.291097e+07,95.776154,81.201000,409.847107,94.849998,4.843087,2010,32.910967,Medium,Medium,Mediun
8,Least developed countries: UN classification,2019,64.972453,1104.186332,1.058014e+09,102.805527,21.900000,41.969666,66.222183,5.101978,2010,1058.013514,Low,Low,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1223,Singapore,2013,82.246341,56967.425794,5.399162e+06,NaN,80.902100,2102.970215,96.550003,4.817631,2010,5.399162,High,High,High
1224,"Iran, Islamic Rep.",2007,73.299000,4687.991663,7.460265e+07,99.747658,9.470000,237.510422,NaN,8.155774,2000,74.602648,Medium,Medium,Mediun
1226,East Asia & Pacific,2013,75.334773,9522.667498,2.267072e+09,104.620399,43.200000,603.489925,94.617699,5.273683,2010,2267.072102,Medium,Medium,Mediun
1227,Antigua and Barbuda,2016,76.968000,16556.855180,8.996900e+04,107.432655,70.974300,817.888916,NaN,4.120368,2010,0.089969,High,High,High


In [137]:
print(f"Average literacy rate for LifeExpectancy > 80: {life_exp['LiteracyRate'].mean()}") 
print(f"Average literacy rate for HealthExpenditure > 8% and GDPGrowth > 3%: {helth_exp['LiteracyRate'].mean()}")

Average literacy rate for LifeExpectancy > 80: 97.19625091552732
Average literacy rate for HealthExpenditure > 8% and GDPGrowth > 3%: 78.27818348572919


## **TASK 9 – NumPy Array Operations**
Using NumPy:
1.	Compute min, max, mean, and range for LifeExpectancy.
2.	Compute 25th, 50th, and 75th percentiles.
3.	Compute correlation between GDPperCapita and LifeExpectancy.


In [152]:
import numpy as np
min_life_exp = np.min(df_sample["LifeExpectancy"]) 
max_life_exp = np.max(df_sample["LifeExpectancy"]) 
mean_life_exp = np.mean(df_sample["LifeExpectancy"]) 
range_life_exp = f"{max_life_exp} - {min_life_exp}"
print(f"Min LifeExpectancy: {min_life_exp}") 
print(f"Max LifeExpectancy: {max_life_exp}") 
print(f"Mean LifeExpectancy: {mean_life_exp}") 
print(f"Range LifeExpectancy: {range_life_exp}")

Min LifeExpectancy: 28.628
Max LifeExpectancy: 85.49634146341464
Mean LifeExpectancy: 68.81137288736028
Range LifeExpectancy: 85.49634146341464 - 28.628


In [139]:
percentiles = np.percentile(df_sample["LifeExpectancy"], q=[25, 50, 75]) 
print(f"25th percentile: {percentiles[0]}") 
print(f"50th percentile: {percentiles[1]}") 
print(f"75th percentile: {percentiles[2]}")

25th percentile: 63.24129238606332
50th percentile: 70.58308207028381
75th percentile: 75.69693595144699


In [140]:
df_clean = df_sample[["GDPperCapita", "LifeExpectancy"]].dropna()
df_clean.shape

(1187, 2)

In [141]:
correlation = np.corrcoef(df_clean["GDPperCapita"], df_clean["LifeExpectancy"])[0, 1] 
print(f"Correlation between GDPperCapita and LifeExpectancy: {correlation}")

Correlation between GDPperCapita and LifeExpectancy: 0.5879829217194033


###  LifeExpectancy Summary 

**Min:** 28.63  
**Max:** 85.50  
**Mean:** 68.81  
**Range:** 85.50 − 28.63 = **56.87**

These results show a **wide variation** in life expectancy across countries, with a gap of almost **57 years** between the lowest and highest values.  
The mean of **~69 years** suggests that most countries fall within a mid-to-high life expectancy range, although the large spread highlights global health disparities.

---

###  LifeExpectancy Percentiles

- **25th percentile:** 63.24  
- **50th percentile (median):** 70.58  
- **75th percentile:** 75.70  

These percentiles indicate that most countries have a life expectancy between **63 and 76 years**, with half of all observations above **70.5**.

---

###  Correlation Interpretation

Correlation values range from **-1 to 1**:

- **-1** → strong negative correlation  
- **0** → no correlation  
- **1** → strong positive correlation  

A correlation **greater than 0.5** indicates a **moderate to strong positive relationship** between the two variables.


In [142]:
# Bonus: Correlation between HealthExpenditure and LifeExpectancy
df_clean = df_sample[["GDPperCapita", "LifeExpectancy", "HealthExpenditure"]].dropna()
print(f"{df_clean.shape}")
correlation = np.corrcoef(df_clean["LifeExpectancy"], df_clean["HealthExpenditure"])[0, 1] 
print(f"Correlation between HealthExpenditure and LifeExpectancy: {correlation}")
correlation = np.corrcoef(df_clean["HealthExpenditure"], df_clean["GDPperCapita"])[0, 1]
print(f"Correlation between HealthExpenditure and GDPperCapita: {correlation}")


(728, 3)
Correlation between HealthExpenditure and LifeExpectancy: 0.582335903054586
Correlation between HealthExpenditure and GDPperCapita: 0.8930604150587699


### Here we can see that exist an even bigger correlation between GDPperCapita and HealthExpenditure than can explain why more rich countries have more LifeExpectancy

## **TASK 10 – Compare 1990 vs 2020**
1.	Identify countries with data for both 1990 and 2020.
2.	Compute change in LifeExpectancy for those countries.
3.	List the top 3 countries with the greatest improvement.


In [143]:
df_1990_2020 = df_sample[df_sample["Year"].isin([1990, 2020])]
df_1990_2020

,Country,Year,LifeExpectancy,GDPperCapita,Population,SchoolEnrollment,InternetUsers,HealthExpenditure,LiteracyRate,GDPGrowth,Decade,PopulationMillions,GDPCategory,GDPCategory_custom,GDPCategory_lambda
16,Norway,1990,76.537317,28242.981403,4.241473e+06,NaN,0.7070,NaN,NaN,1.932291,1990,4.241473,High,High,High
21,OECD members,1990,74.538567,17011.413527,1.107696e+09,NaN,NaN,NaN,NaN,2.937451,1990,1107.695519,High,High,High
37,Kenya,1990,58.334000,374.458993,2.289265e+07,105.313667,0.0000,NaN,NaN,4.192051,1990,22.892651,Low,Low,Low
43,Myanmar,2020,66.612000,1490.216835,5.301652e+07,NaN,45.4477,66.022842,93.547129,-9.048347,2020,53.016522,Low,Low,Low
46,Israel,1990,76.607317,13308.311060,4.660000e+06,NaN,0.1110,NaN,NaN,7.315199,1990,4.660000,Medium,Medium,Mediun
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1194,Equatorial Guinea,2020,62.141000,5764.055029,1.716468e+06,NaN,55.0013,230.321442,NaN,-4.791081,2020,1.716468,Medium,Medium,Mediun
1195,"Congo, Dem. Rep.",1990,48.538000,254.872656,3.668406e+07,59.328461,0.0000,NaN,NaN,-6.568311,1990,36.684063,Low,Low,Low
1206,Chad,1990,45.074000,290.599045,5.982833e+06,50.201832,0.0000,NaN,NaN,-4.178472,1990,5.982833,Low,Low,Low
1222,Morocco,2020,73.133000,3268.030273,3.658421e+07,112.354037,84.1204,185.452454,NaN,-7.178207,2020,36.584208,Low,Low,Low


In [144]:
boths_years = df_1990_2020.pivot(index="Country", columns="Year", values="LifeExpectancy").dropna()
boths_years

Year,1990,2020
Country,,
Brunei Darussalam,71.629000,75.114000
Bulgaria,71.641463,73.356098
Chad,45.074000,53.082000
Chile,73.675000,79.349000
"Congo, Dem. Rep.",48.538000,60.428000
"Iran, Islamic Rep.",64.066000,74.141000
Latvia,69.273171,75.185366
World,65.107983,72.182224



To compare changes in **LifeExpectancy** between 1990 and 2020, we pivoted the dataset so that **countries become the index**, **years become columns**, and **LifeExpectancy** becomes the cell values.  
This structure makes it easy to directly calculate how much life expectancy increased over the 30-year period.

After filtering only the countries that have data for **both 1990 and 2020**, we compute the difference:


In [145]:
change_life_exp = boths_years[2020] - boths_years[1990] 
change_life_exp

Country
Brunei Darussalam      3.485000
Bulgaria               1.714634
Chad                   8.008000
Chile                  5.674000
Congo, Dem. Rep.      11.890000
Iran, Islamic Rep.    10.075000
Latvia                 5.912195
World                  7.074241
dtype: float64

In [146]:
bigest_change = change_life_exp.nlargest(3) 
bigest_change

Country
Congo, Dem. Rep.      11.890
Iran, Islamic Rep.    10.075
Chad                   8.008
dtype: float64

In [147]:
bigest_change_v2 = change_life_exp.sort_values(ascending=False).head(3) 
bigest_change_v2

Country
Congo, Dem. Rep.      11.890
Iran, Islamic Rep.    10.075
Chad                   8.008
dtype: float64

The countries with the **greatest improvement** in life expectancy between 1990 and 2020 are:

- **Congo, Dem. Rep.**  
- **Iran, Islamic Rep.**  
- **Chad**  

These countries show the largest positive changes, indicating significant progress in health conditions, healthcare access, and overall living standards over the past decades.